_posts/2023-10-11-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,485 @@
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetCommitCallback;
+import org.apache.kafka.clients.consumer.OffsetOutOfRangeException;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.RebalanceInProgressException;
+import org.apache.kafka.common.errors.RetriableException;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.LongDeserializer;
+
+import static java.time.Duration.ofMillis;
+import static java.util.Collections.singleton;
+
+public class Consumer implements ConsumerRebalanceListener, OffsetCommitCallback {
+    // Constants for configuration
+    private static final String BOOTSTRAP_SERVERS = "localhost:9092";
+    private static final String GROUP_ID = "my-group";
+    private static final long POLL_TIMEOUT_MS = 1_000L;
+    private static final String TOPIC_NAME = "my-topic";
+    private static final long NUM_MESSAGES = 50;
+    private static final long PROCESSING_DELAY_MS = 1_000L;
+
+    private KafkaConsumer<Long, byte[]> kafkaConsumer;
+    protected AtomicLong messageCount = new AtomicLong(0);
+    private Map<TopicPartition, OffsetAndMetadata> pendingOffsets = new HashMap<>();
+
+    public static void main(String[] args) {
+        new Consumer().run();
+    }
+
+    public void run() {
+        System.out.println("Running consumer");
+
+        // Create a Kafka consumer instance
+        // This consumer receives messages from the Kafka topic asynchronously
+        try (var consumer = createKafkaConsumer()) {
+            kafkaConsumer = consumer;
+            consumer.subscribe(singleton(TOPIC_NAME), this);
+            System.out.printf("Subscribed to %s%n", TOPIC_NAME);
+            while (messageCount.get() < NUM_MESSAGES) {
+                try {
+                    // Poll for new records from Kafka
+                    ConsumerRecords<Long, byte[]> records = consumer.poll(ofMillis(POLL_TIMEOUT_MS));
+                    if (!records.isEmpty()) {
+                        for (ConsumerRecord<Long, byte[]> record : records) {
+                            System.out.printf("Record fetched from %s-%d with offset %d%n",
+                                record.topic(), record.partition(), record.offset());
+                            sleep(PROCESSING_DELAY_MS);
+
+                            // Track pending offsets for commit
+                            pendingOffsets.put(new TopicPartition(record.topic(), record.partition()),
+                                new OffsetAndMetadata(record.offset() + 1, null));
+                            if (messageCount.incrementAndGet() == NUM_MESSAGES) {
+                                break;
+                            }
+                        }
+                        // Commit pending offsets asynchronously
+                        consumer.commitAsync(pendingOffsets, this);
```

Comment:
@fvaleri -- what do you think of this suggestion?

Reply:
We can't get rid of pendingOffsets, because it is used in onPartitionsRevoked. We can use commitAsync(this) on this line, but I would prefer to keep it as it is. Already discussed here: https://github.com/strimzi/strimzi.github.io/pull/386#discussion_r1332871834.